lib/hipcore/patterns.ml

```diff
@@ -0,0 +1,287 @@
+open Typedhip
+open Common
+open Syntax.Typed
+
+type guarded_pattern = pattern * term
+
+let make pat typ = {
+  pattern_desc = pat;
+  pattern_type = typ
+}
+
+let alias pat name = make (PAlias (pat, name)) pat.pattern_type
+
+let (let*) xs f = List.concat_map f xs
+let return x = [x]
+
+let cartesian_product (lists: 'a list list) : 'a list list =
+  List.fold_right (fun ls result -> let* x = ls in let* rest = result in return (x::rest)) lists [[]]
```

Comment:
For consistency, this should be eventually moved to `Utils.Lists`.

Reply:
Fixed in [d45dfa2](https://github.com/hipsleek/Heifer/pull/34/commits/d45dfa2415bf08595161a05cd6d79df51dd0b687).